apps/src/lib/node/ledger/ethereum_node/mod.rs

```diff
@@ -209,22 +208,19 @@ pub mod eth_fullnode {
         /// received from the Oracle process. If either, return the
         /// status.
         async fn wait(&mut self) -> Result<()> {
-            loop {
-                match self.process.try_wait() {
-                    Ok(Some(status)) => {
-                        return if status.success() {
-                            Ok(())
-                        } else {
-                            Err(Error::Runtime(status.to_string()))
-                        };
-                    }
-                    Ok(None) => {}
-                    Err(err) => return Err(Error::Runtime(err.to_string())),
+            use futures::future::{self, Either};
+
+            let child_proc = self.process.wait();
+            futures::pin_mut!(child_proc);
```

Comment:
How come we have to pin the future?

Reply:
because the future we return from `let child_proc = self.process.wait();` is not [`Unpin`](https://doc.rust-lang.org/std/marker/trait.Unpin.html). it fails to compile if we don't pin it to the stack.

Reply:
```
error[E0277]: `from_generator::GenFuture<[static generator@tokio::process::Child::wait::{closure#0}]>` cannot be unpinned
    --> apps/src/lib/node/ledger/ethereum_node/mod.rs:216:56
     |
216  |             match future::select(&mut self.abort_recv, child_proc).await {
     |                   --------------                       ^^^^^^^^^^ within `impl futures::Future<Output = std::result::Result<ExitStatus, std::io::Error>>`, the trait `Unpin` is not implemented for `from_generator::GenFuture<[static generator@tokio::process::Child::wait::{closure#0}]>`
     |                   |
     |                   required by a bound introduced by this call
     |
    ::: /home/sugo/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/process/mod.rs:1116:37
```

Reply:
@james-chf basically this just means that if you `poll()` a future and return `Poll::Pending`, you can't guarantee you'll access the same value on a later call to `poll()` unless it is `Pin`. simultaneously, you can move values out of a `Pin<T>` if `T: Unpin`.